File: grind75/Valid-Parentheses.py
#straitforward. Uses a dictionary and a stack.
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        
        paren_pairs = {'}':'{' ,']': '[',')': '('}
        for char in s:
            if char in "([{":
                stack.append(char)
            elif char in "}])":
                if not stack or stack.pop() != paren_pairs[char]:
                    return False
        if not stack:
            return True
        else:
            return False
               

File: easy-leetcode-solutions/cpp-solutions/valid-parentheses.cpp
#include <stack>
#include <string>
class Solution {
public:
    bool isValid(std::string s) {
        std::stack<char> string_stack;

        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {
                string_stack.push(c);
            } 
            else {
                if (string_stack.empty()) return false;
                char top = string_stack.top();
                string_stack.pop();

                if ((c == ')' && top != '(') ||
                    (c == '}' && top != '{') ||
                    (c == ']' && top != '[')) {
                    return false;
                }
            }
        }
        return string_stack.empty();
    }
};
